enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,249 @@
+
+## Motivation
+
+Having two different provisioning flows comes with significant maintenance and
+development costs:
+- Some features, namely firmware and RAID configuration, require using the
+  Ironic agent, which is not present in the *live ISO* flow.
+- Supporting (i)PXE boot with the *live ISO* flow requires creating new API
+  in baremetal-operator specifically for this purpose.
+- In the *live ISO* flow, it is not possible to notify Ironic (and thus
+  baremetal-operator) when the actual installation finishes. From the
+  perspective of Metal3 components, the installation is done once the ISO is
+  connected. As a result, Ironic cannot disconnect the ISO from the machine's
+  BMC, nor can it update the BMC's boot device settings properly.
+- The *live ISO* flow disables Ironic *inspection* process and has to simulate
+  it by updating the `BareMetalHost` based on the assisted service's inspection
+  process. This creates an unnessesary coupling between the assisted service
+  and the baremetal-operator's hardware inventory format (which, in turn, is
+  derived from the Ironic's inventory format).
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Making assisted components mandatory for all bare-metal deployments.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). It will be updated to use the
+  assisted service's `InfraEnv` resources, when they are present for the given
+  `BareMetalHost`.
+- When `InfraEnv` resources are present, the generated images will contain
+  Ignition configuration for both agents. The Ironic agent will be started
+  automatically and handle coordination with Ironic and baremetal-operator.
+  The configuration will also supress automatic start-up of the assisted
+  agent.
+- [Ironic agent][ironic-agent] currently contains a custom deployment plugin
+  (*custom deploy method* in baremetal-operator terms, *custom deploy step* in
+  Ironic terms) that invokes `coreos-installer` for the image installation.
+  A new custom deploy method `install_assisted` will be added to start the
+  assisted agent at the right moment.
+- The assisted service will stop explicitly disabling *inspection* on
+  `BareMetalHost`s it manages, will start using a different process based
+  on the  new `install_assisted` deploy method and stop rebooting the machine
+  in the end of the process.
+
+### User Stories
+
+N/A (this change itself does not add new user stories)
```

Comment:
Users wanting firmware/RAID with ZTP?

Reply:
+1 we could also perhaps have stories around environments where the in-band efibootmgr approach doesn't work reliably, which would be solved by having Ironic managed the ISO attach/detach like in the regular virtualmedia flow.

Another story would be PXE boot for ZTP customers (which could be solved without doing this, but that becomes much easier as a side-effect of this alignment?)